src/jsglue.cpp

```diff
@@ -31,6 +31,50 @@ typedef size_t(*GetSize)(JSObject *obj);
 
 WantToMeasure gWantToMeasure = nullptr;
 
+struct JobQueueTraps {
+  JSObject* (*getIncumbentGlobal)(void* queue, JSContext* cx);
+  bool (*enqueuePromiseJob)(void* queue, JSContext* cx, JS::HandleObject promise,
+                            JS::HandleObject job, JS::HandleObject allocationSite,
+                            JS::HandleObject incumbentGlobal) = 0;
+  bool (*empty)(void* queue);
+};
+
+class RustJobQueue : public JS::JobQueue{
+  JobQueueTraps mTraps;
+  void* mQueue;
+public:
+  RustJobQueue(const JobQueueTraps& aTraps, void* aQueue)
+  : mTraps(aTraps)
+  , mQueue(aQueue)
+  {
+  }
+
+  virtual JSObject* getIncumbentGlobal(JSContext* cx) {
+    return mTraps.getIncumbentGlobal(mQueue, cx);
+  }
+
+  bool enqueuePromiseJob(JSContext* cx, JS::HandleObject promise,
+                         JS::HandleObject job, JS::HandleObject allocationSite,
+                         JS::HandleObject incumbentGlobal)
+  {
+    return mTraps.enqueuePromiseJob(mQueue, cx, promise, job, allocationSite, incumbentGlobal);
+  }
+
+  virtual bool empty() const {
+    return mTraps.empty(mQueue);
+  }
+
+  virtual void runJobs(JSContext* cx) {
+    MOZ_ASSERT("runJobs should not be invoked");
```

Comment:
Hmm, this is a bit annoying, I can imagine it being triggered in future versions of SM/Servo.